internal/flightplan/matrix_decoder.go

```diff
@@ -0,0 +1,272 @@
+			Subject:  block.TypeRange.Ptr(),
+			Context:  block.DefRange.Ptr(),
+		})
+	}
+
+	// Let's decode our matrix block into a matrix
+	block = mBlocks[0]
+	matrix := NewMatrix()
+	res := &DecodedMatrices{Original: matrix}
+
+	// Each attribute in the matrix should be a variant name whose value must
+	// be a list of strings. Convert the value into a matrix vector and add it.
+	// We're ignoring the diagnostics JustAttributes() will return here because
+	// there might also be include and exclude blocks.
+	mAttrs, _ := block.Body.JustAttributes()
+	for _, attr := range md.sortAttributes(mAttrs) {
+		vec, moreDiags := md.decodeMatrixAttribute(ctx, block, attr)
+		diags = diags.Extend(moreDiags)
+		if moreDiags != nil && moreDiags.HasErrors() {
+			continue
+		}
+
+		res.Original.AddVector(vec)
+	}
+
+	// Now that we have our basic variant vectors in our matrix, we need to combine
+	// all vectors into a product that matches all possible unique value combinations.
+	res.FinalProduct = res.Original.CartesianProduct().UniqueValues()
+
+	// Now we need to go through all of our blocks and process include and exclude
+	// directives. Since HCL allows us to use ordering we'll apply them in the
+	// order in which they're defined.
+	blockC, remain, moreDiags := block.Body.PartialContent(&hcl.BodySchema{
+		Blocks: []hcl.BlockHeaderSchema{
+			{Type: blockTypeMatrixInclude},
+			{Type: blockTypeMatrixExclude},
+		},
+	})
+	diags = diags.Extend(moreDiags)
+	if moreDiags != nil && moreDiags.HasErrors() {
+		return nil, diags
+	}
+	diags = diags.Extend(verifyBodyOnlyHasBlocksWithLabels(
+		remain, blockTypeMatrixInclude, blockTypeMatrixExclude,
+	))
+
+	for _, mBlock := range blockC.Blocks {
+		switch mBlock.Type {
+		case "include":
+			iMatrix := NewMatrix()
+			iAttrs, moreDiags := mBlock.Body.JustAttributes()
+			diags = diags.Extend(moreDiags)
+			if moreDiags != nil && moreDiags.HasErrors() {
+				continue
+			}
+
+			for _, attr := range md.sortAttributes(iAttrs) {
+				vec, moreDiags := md.decodeMatrixAttribute(ctx, mBlock, attr)
+				diags = diags.Extend(moreDiags)
+				if moreDiags != nil && moreDiags.HasErrors() {
+					continue
+				}
+
+				iMatrix.AddVector(vec)
+			}
+
+			// Generate our possible include vectors and add them to our main matrix.
+			iMatrixProduct := iMatrix.CartesianProduct().UniqueValues()
+			res.IncludeProducts = append(res.IncludeProducts, iMatrixProduct)
+
+			for _, vec := range iMatrixProduct.GetVectors() {
+				res.FinalProduct.AddVector(vec)
+			}
+		case "exclude":
+			eMatrix := NewMatrix()
+			eAttrs, moreDiags := mBlock.Body.JustAttributes()
+			diags = diags.Extend(moreDiags)
+			if moreDiags != nil && moreDiags.HasErrors() {
+				continue
+			}
+
+			for _, attr := range md.sortAttributes(eAttrs) {
+				vec, moreDiags := md.decodeMatrixAttribute(ctx, mBlock, attr)
+				diags = diags.Extend(moreDiags)
+				if moreDiags != nil && moreDiags.HasErrors() {
+					continue
+				}
+				eMatrix.AddVector(vec)
+			}
```

Comment:
Nit: this code appears to be verbatim of the include case. Wonder if this could be shared logic.

Reply:
Refactored all matrix attribute decoding into a function which we now use at both top level keys and in `include` and `exclude` blocks.